File: basic/dfsbfs/treeproblem/LeafNodeCounter.java
// problem link : https://www.acmicpc.net/problem/1068
// [solution]
// DFS를 사용하여 트리를 탐색합니다.
// 각 노드에서 자식이 있는지 확인하고, 삭제된 노드는 탐색하지 않고 
// 자식이 없다면 리프 노드로 카운트합니다.

// 시간 복잡도는 O(N)

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;
import java.util.Queue;
import java.util.StringTokenizer;

public class LeafNodeCounter {
    static List<Integer>[] adj;
    static boolean[] visited;
    static int rootNode;
    static int deleteNode;
    static int leafCount;

    public static void dfs(int node) {
        // 현재 노드를 방문 처리
        visited[node] = true;
        
        boolean hasChild = false;
        // 현재 노드와 연결된 모든 인접 노드를 탐색
        for (int next : adj[node]) {
            if (next == deleteNode) {
                continue;
            }
            // 인접 노드가 아직 방문되지 않았다면
            if (!visited[next]) {
                // 인접 노드에서 DFS를 재귀적으로 호출하여 탐색 계속
                hasChild = true;
                dfs(next);
            }
        }
        if (!hasChild) {
            leafCount++;
        }
    }

    static void bfs(int start) {
        Queue<Integer> queue = new LinkedList<>();

        queue.add(start);
        visited[start] = true;

        // 큐가 빌 때까지 반복합니다.
        while(!queue.isEmpty()) {
            // 큐에서 노드를 꺼내고, 
            int node = queue.poll();

            boolean hasChild = false;
            // 해당 노드의 인접 노드 중 방문하지 않은 노드를 큐에 삽입합니다.
            for (int next : adj[node]) {
                if (next == deleteNode) {
                    continue;
                }
                if (!visited[next]) {
                    visited[next] = true;
                    hasChild = true;
                    queue.add(next);
                }
            }
            if (!hasChild) {
                leafCount++;
            }
        }
    }

    public static void main(String[] args) throws IOException {
        // 빠른 입력을 위한 BufferedReader 사용
        BufferedReader br = new BufferedReader(
            new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        int n = Integer.parseInt(st.nextToken()); // 정점 개수

        // 인접리스트 초기화
        adj = new ArrayList[n];
        for (int i = 0; i < n; i++) {
            adj[i] = new ArrayList<>();
        }
        st = new StringTokenizer(br.readLine());
        for (int i = 0; i < n; i++) {
            int p = Integer.parseInt(st.nextToken());
            if (p == -1) {
                rootNode = i;
                continue;
            }
            adj[p].add(i);
        }

        deleteNode = Integer.parseInt(br.readLine());

        if (deleteNode == rootNode) {
            System.out.println(0);
            return;
        }

        visited = new boolean[n + 1];
        leafCount = 0;

        //dfs(rootNode);
        bfs(rootNode);

        System.out.println(leafCount);
    }
}

File: basic/dfsbfs/treeproblem/LeafNodeCounter.py
import sys
from collections import deque

input = sys.stdin.readline

# 전역 변수
adj = []
visited = []
delete_node = 0
leaf_count = 0

#dfs 구현
def dfs(node):
    global leaf_count, adj, visited, delete_node
    visited[node] = True
    has_child = False
    
    for next_node in adj[node]:
        # 삭제된 노드는 건너뛰기
        if next_node == delete_node:
            continue
        
        # 아직 방문하지 않은 노드가 있다면
        if not visited[next_node]:
            has_child = True
            dfs(next_node)
    
    # 자식 노드가 없다면 리프 노드
    if not has_child:
        leaf_count += 1

#bfs 구현
def bfs(start):
    global leaf_count, adj, visited, delete_node

    queue = deque([start])
    visited[start] = True
    while queue:
        node = queue.popleft()
        has_child = False

        for next in adj[node]:
            # 삭제된 노드는 건너뛰기
            if next == delete_node:
                continue
            if (not visited[next]):
                has_child = True
                visited[next] = True
                queue.append(next)

        # 자식 노드가 없다면 리프 노드
        if not has_child:
            leaf_count += 1

# 노드의 개수 입력
n = int(input())

# 트리 초기화 (인접 리스트)
adj = [[] for _ in range(n)]

# 부모 노드 정보 입력
parents = list(map(int, input().split()))
root_node = -1

# 트리 구성
for i in range(n):
    parent = parents[i]
    if parent == -1:
        root_node = i  # 루트 노드 기록
    else:
        adj[parent].append(i)  # 부모 노드에 자식 추가
        adj[i].append(parent)  # 양방향으로 연결 (트리 탐색용)

# 삭제할 노드 입력
delete_node = int(input())

# 방문 배열 초기화
visited = [False] * n
leaf_count = 0

# 삭제할 노드가 루트 노드인 경우, 트리가 없어지므로 리프 노드는 0개
if delete_node == root_node:
    print(0)
else:
    #dfs(root_node)
    bfs(root_node)
    print(leaf_count)